packages/@react-spectrum/s2/src/SegmentedControl.tsx

```diff
@@ -238,10 +252,10 @@ function SegmentedControlItem(props: ControlItemProps, ref: FocusableRef<HTMLLab
       ref={domRef} 
       inputRef={inputRef}
       style={props.UNSAFE_style}
-      className={renderProps => (props.UNSAFE_className || '') + controlItem({...renderProps}, props.styles)} >
+      className={renderProps => (props.UNSAFE_className || '') + controlItem({...renderProps, isIconOnly: isIconOnly?.current}, props.styles)} >
```

Comment:
can't read from ref during render, you'll need to make this state and trigger a re-render
you have it in a layouteffect, so this will update fast enough that there won't be a flash of wrong spacing